### PR TITLE
Call socialdb to notify

### PR DESCRIPTION
**Pull request by ailisp**
This PR implements the contract part of Dev Gigs Board notification feature. It is currently deployed on `devgigsboard.bo.near`. The UI part is implemented and tested in https://near.social/#/devgigsboard.bo.near/widget/Ideas and https://near.social/#/devgigsboard.bo.near/widget/NotificationFeed.

# My test steps:
1. create two accounts, `alice.bo.near`, `bob.bo.near`, register in near.social, **go to the profile page, gives each of them a name**  (e.g. "Test Alice" / Test Bob") **!!!important**.
2. Sign in with `bob.bo.near`. Go to main board: https://near.social/#/devgigsboard.bo.near/widget/Ideas. The first time sign in to this fork's board, it will ask you to sign a transaction to allow `devgigsboard.bo.near` to set your `index/notify` key in social db. This is required because social db's permission model. Otherwise `devgigsboard.bo.near` cannot write state under your social db root key, and cannot trigger notification. It looks like this (screenshot is taken for `carl.bo.near`, bob's looks all same but arguments contains bob instead):
image

3. Submit an idea: https://near.social/#/devgigsboard.bo.near/widget/Ideas?postId=3
4. Sign out, sign in with Alice, like and comment to the idea.
5. Sign out, sign in with `devgigsboard.bo.near`, edit to the idea.
6. Sign out, sign in with Bob, observe there're 3 notifications. Go to https://near.social/#/devgigsboard.bo.near/widget/NotificationFeed you can see:

image

Note that if you click to the notification bell, it will leads you to the original NotificationFeed widget instead of my fork, which doesn't recognize Dev Gigs type of notification:

image

I'll ask @evgenykuzyakov if he is okay to merge my fork.

### CODE REVIEW

**Comment by maxzaver** [comment]
Okay, so I guess we need to wait for @evgenykuzyakov to merge the code into near.social before we enable it on DevGigs board. @ailisp could you handle the communication with him?

File: src/lib.rs
Comment (maxzaver) [inline src/lib.rs]: If someone passes an existing `parent_id` it should always be the case that `post_to_children` contains the record for it, unless `parent_id==ROOT_POST_ID`. What is the use of this change?

Comment (maxzaver) [reply src/lib.rs]: If the intent to make it work with `parent_id ==ROOT_POST_ID` then we can just do an early return of `None` here.

File: src/lib.rs
Comment (maxzaver) [inline src/lib.rs]: I am not sure we need to wait for `notify::notify_reply`, which is what would happen if we return `Promise` instead of just dropping it upon creation.

File: src/lib.rs
Comment (maxzaver) [inline src/lib.rs]: I think https://docs.rs/near-sdk/latest/near_sdk/enum.PromiseOrValue.html is more correct here.

File: src/lib.rs
Comment (maxzaver) [inline src/lib.rs]: Same thing, I don't think we need to wait on the other contract to execute, so we can probably drop the promise without returning it.

File: src/lib.rs
Comment (maxzaver) [inline src/lib.rs]: Same thing, I don't think we need to wait on the other contract to execute, so we can probably drop the promise without returning it.

**Review by maxzaver — CHANGES_REQUESTED** [review]

File: src/lib.rs
Comment (ailisp) [reply src/lib.rs]: **I'll revert this line, let's address this problem in a separate PR.**

> If the intent to make it work with parent_id ==ROOT_POST_ID 

Yes this was intend to bootstrap the state - where I first deploy contract and create first post it panics here because initially root id doesn't have siblings (`post_to_children.get(ROOT_ID)` is `None`).

> then we can just do an early return of None here.

Do you mean something like:
```
if parent_id != ROOT_POST_ID {
...
} else {
// doesn't do anything with top level post's sibling, aka ROOT_POST's children.
  return None
}
 ```

But in `main` branch, after insert an top-level post, this line:
```
        self.post_to_children.insert(&parent_id, &siblings);
```
will insert/update an entry `ROOT_ID` -> `vector of all top level posts`. Do we want to keep this entry?

File: src/lib.rs
Comment (ailisp) [reply src/lib.rs]: That's good to know! Not waiting for notify is better, so I'll change to drop after creation.

File: src/lib.rs
Comment (ailisp) [reply src/lib.rs]: Because of above comment this function will return unit. 

Regarding `PromiseOrValue<()>` vs `Option<Promise>` problem, I thought the later is fine because I observed it in non_fungible_token standard: https://github.com/near/near-sdk-rs/blob/770cbce018a1b6c49d58276a075ace3da96d6dc1/near-contract-standards/src/non_fungible_token/approval/approval_impl.rs#L29

File: src/lib.rs
Comment (ailisp) [reply src/lib.rs]: Sounds good!

File: src/lib.rs
Comment (ailisp) [reply src/lib.rs]: OK!

**Comment by ailisp** [comment]
@nearmax 
> Okay, so I guess we need to wait for @evgenykuzyakov to merge the code into near.social before we enable it on DevGigs board. @ailisp could you handle the communication with him?

Of course! Before I come to Evgeny, let's confirm my fork of notification feed widget is final. Do you have any tweaks in mind / additional information to show on the UI?

File: src/lib.rs
Comment (nearmax) [reply src/lib.rs]: I think then NFT standard has a bug. I am not sure returning anything as a type besides `PromiseOrValue`, `Promise` or regular value works. CC @frol

Comment (maxzaver) [reply src/lib.rs]: I think then NFT standard has a bug. I am not sure returning anything as a type besides `PromiseOrValue`, `Promise` or regular value works. CC @frol

File: src/lib.rs
Comment (maxzaver) [reply src/lib.rs]: The current code of this PR looks good.

**Review by maxzaver — APPROVED** [review]
I did not get a chance to test the UI itself, because it contains lots of steps. However, I assume when clicking on notification in leads to an actual post. Perhaps we could rename "replied/edited/liked your Developer Gigs post" to "replied/edited/liked your Developer Governance post".

Please don't forget to rebuild wasm file using recent main branch

File: src/lib.rs
Comment (frol) [reply src/lib.rs]: I also believe `PromiseOrValue` should be the return type. I don't know why `nft_approve` has this signature, but I can dive into it if necessary.

**Comment by ailisp** [comment]
> I assume when clicking on notification in leads to an actual post. 

Yes. This is tested.

> Perhaps we could rename "replied/edited/liked your Developer Gigs post" to "replied/edited/liked your Developer Governance post".

OK.

> Please don't forget to rebuild wasm file using recent main branch

Double checked, current wasm is generated from: merge current near/main (at the time of writing this comment) to this branch, `./build.sh`. This wasm is also deployed and tested again with UI.

After merge this PR, `./build.sh` from near/main also doesn't change `near_ideas.wasm`.

**Pull request merged**
